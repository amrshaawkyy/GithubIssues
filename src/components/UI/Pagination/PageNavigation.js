import React, { Component } from 'react';
import { Swipeable } from 'react-swipeable'
class PageNavigation extends Component {
    render() {
        return (
            <Swipeable
                onSwipedLeft={this.props.onSwipingLeft}
                onSwipedRight={this.props.onSwipingRight}
            >
                {this.props.children}
            </Swipeable>
        )
    }
}
export default PageNavigation;