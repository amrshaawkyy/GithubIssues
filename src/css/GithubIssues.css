.navbar-center {
    display: flex;
    align-items: center;
    flex-direction: row;
    flex-wrap: wrap;
}
.navbar-center >p {
    margin-left: 15px;
}

/* .navbar.is-fixed-bottom, .navbar.is-fixed-top {
    width: 100%;
	position: fixed;
	bottom: 0;
	z-index: 1;
    background-color: red;
} */

.rotate {
    -moz-transition: all 2s linear;
    -webkit-transition: all 2s linear;
    transition: all 2s linear;
}

.DateRangePickerInput {
    border-radius: 10px;
    width: 70%;
    cursor: pointer;
    height: 40px;
    text-align: center;
    border: 1px solid lightgray;
    font-weight: bold;
    color: dimgrey;
}
.DateRangePickerInput:hover {
    background-color: lightgray;
}

.rotate.down {
    -moz-transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    transform: rotate(90deg);
}

.panel-icon {
    line-height: 0.5rem;
    text-align: center;
    vertical-align: middle;
    font-size: 10px;
    padding-top: 10%;
    padding-left: 10%;
}

/* .pagination{
    background: transparent;
} */

.pagination>button, button.pagination-link {
    background-color: #ffffff;
    border-radius: 50%;
}

/* .custom-section-1 {
    padding-top: 12px !important;
    padding-right: 0.75em;
    padding-bottom: 0.5em;
    padding-left: 0.75em;
} */

/* 
.fa-check {
    font-size: 100%;
    margin: -0.01rem 0 0 0.9rem;
} */

.tag {
    margin: 2px;
    height: 1.2em !important;
}

.time {
    font-size: 0.8rem;
    width: 100%;
}

.bodyContainer {
    max-height: fit-content;
    display: block;
    background-color: #ffffff;
    margin-top: 10px;
    overflow-wrap: normal;
    overflow: hidden;
    padding: 10px !important;
}

.bodyContainer>li, ol, pre, textarea, ul {
    padding: inherit;
    /* margin: 10px !important; */
}

/* .panel-heading {
    padding-top: 0.5em;
    padding-right: 0.75em;
    padding-bottom: 0em;
    padding-left: 0.75em;
} */

.panel-block, .panel-heading, .panel-tabs {
    border-top: 1px solid #dbdbdb;
    border-bottom: 1px solid #dbdbdb;
    border-left: 1px solid #dbdbdb;
    border-right: 1px solid #dbdbdb;
}

.panel-block {
    flex-flow: wrap;
    align-items: center;
    color: #363636;
    padding: .5em .75em;
    width: 100%;
    margin: 0px !important;
    cursor: pointer;
}

.panel-ovr {
    border: none !important;
    margin: 0px !important;
    width: 100%;
    max-width: 100%;
    line-height: 1.25;
}

._title {
    text-overflow: ellipsis;
    overflow-x: hidden;
    max-width: 90%;
    overflow-wrap: normal;
    color: #363636 !important;
    margin-right: 12px;
    font-weight: 900;
}

.milestone, .comments {
    margin-left: 20px;
}

.show-body {
    padding: 5px;
}

.panel-icon-container {
    margin-right: -3.5% !important;
}

.info-container {
    margin-left: -2vw;
    margin-top: 2vw;
    /*check*/
    line-height: 1;
}

@media screen and (max-width: 769px) {
    .panel-icon-container {
        margin-right: 0 !important;
    }
    .info-container {
        margin-left: 0% !important;
        padding-left: 40px;
    }
    .milestone {
        margin-left: 0px !important;
    }
    .title-string {
        font-size: 4vw;
    }
}