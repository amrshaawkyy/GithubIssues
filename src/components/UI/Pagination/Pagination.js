import React from 'react';
import PageNavigation from './PageNavigation';
import Arrow from '../Arrow/Arrow'
import PropTypes from 'prop-types';
const pagination = (props) => {
    const sendPageNumber = (e) => {
        props.onPageChange(`${e.currentTarget.dataset.url}`);
    }
    const swipingLeft = (e, absX) => {
        const lastpage = props.pages.last ? props.pages.last.split("=").pop() : props.pages.prev ? +props.pages.prev.split("=").pop() + 1 : 1
        if (props.currentPageNumber < lastpage) {
            props.onPageChange(`&page=${props.currentPageNumber + 1}`)
        };
    }
    const swipingRight = (e, absX) => {
        if (props.currentPageNumber > 1) {
            props.onPageChange(`&page=${props.currentPageNumber - 1}`)
        };
    }
    const pages = props.pages;
    const currentPageNumber = props.currentPageNumber
    const lastpage = pages.last ? pages.last.split("=").pop() : pages.prev ? +pages.prev.split("=").pop() + 1 : 1
    const pageButtons = Array.apply(null, { length: lastpage })
        .map(Number.call, Number)
        .map(x => `&page=${x + 1}`)
    let button = null;
    if (pageButtons.length > 12) {
        button = <button
             className=" pagination-link  is-hidden-mobile">{`${currentPageNumber}/${lastpage}`}
        </button>
    }
    else {
        button = pageButtons.map((page, i) => {
            const p = +page.split("=").pop()
            return <li key={i} className="is-hidden-mobile">
                <button
                    data-url={page}
                    onClick={(event) => sendPageNumber(event)}
                    className={"pagination-link " + (p === (currentPageNumber) ? " is-current" : p)}
                    aria-label={"Goto page " + p} >{p}
                </button>
            </li>
        })
    }
    return (
        <PageNavigation onSwipingLeft={(event) => swipingLeft(event)} onSwipingRight={(event) => swipingRight(event)}>
            <nav className="pagination is-small is-centered" role="navigation" aria-label="pagination">
                <button
                    className={"pagination-previous " + (currentPageNumber > 1 ? "" : " is-invisible")}
                    data-url={`&page=${currentPageNumber - 1}`}
                    onClick={(event) => sendPageNumber(event)}>
                    <Arrow angle={270} length={12} width={30} filled={false} color="#80868B" />
                </button>
                <button
                    className="pagination-link is-hidden-tablet">{`${currentPageNumber}/${lastpage}`}
                </button>
                <button
                    className={"pagination-next " + (currentPageNumber < lastpage ? "" : " is-invisible")}
                    data-url={`&page=${currentPageNumber + 1}`}
                    onClick={(event) => sendPageNumber(event)}>
                    <Arrow angle={90} length={12} width={30} filled={false} color="#80868B" />
                </button>
                <ul className="pagination-list">
                    {button}
                </ul>
            </nav>
        </PageNavigation >
    );
}
pagination.propTypes = {
    onPageChange: PropTypes.func,
    pages: PropTypes.object,
    currentPageNumber: PropTypes.number
};
export default pagination;