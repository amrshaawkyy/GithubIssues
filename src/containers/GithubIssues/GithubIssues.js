import React, { Component } from 'react';
import IssuesFiltering from '../IssuesFiltering/IssuesFiltering';
import Pagination from '../../components/UI/Pagination/Pagination';
import 'bulma/css/bulma.css';
import '../../css/GithubIssues.css';
import Spinner from '../../components/UI/Spinner/Spinner';
import GitHubIssue from '../../components/GitHubIssue/GitHubIssue';
class GithubIssues extends Component {
    state = {
        user: 'facebook',
        repo: 'react',
        listFilter: { state: "open", choice: "issues" },
        since: new Date('01/01/'+ new Date().getFullYear()),
        per_page: 40,
        sort: true,
        page: '',
        currentPageNumber: 1,
        pages: {},
        issues: [],
        loading: true,
        error: null,
        showBody: {},
        sortName: "Latest"
    }
    componentDidMount() {
        this.fetchAllIssues();
    }
    shouldComponentUpdate(previousState, nextState) {
        if (previousState.page !== nextState.page) {
            return true;
        }
        if (previousState.listFilter !== nextState.listFilter) {
            return true;
        }
        return false;
    }
    fetchAllIssues() {
        const headers = {
            headers: {
                Authorization: "token 60e07ff280c786e523a632be9af8f992270a5c5b",
                Accept: "application/vnd.github.v3+json,application/vnd.github.machine-man-preview+json",
            }
        };
        const baseUrl = "https://api.github.com/repos"
        let url = { state: this.state.listFilter.state, since: this.state.since, per_page: 40, sort: this.state.sort ? "created" : "created-asc" };
        const keys = Object.keys(url)
        const params = keys.length ?
            "?" + keys
                .map(key => encodeURIComponent(key) +
                    "=" + encodeURIComponent(url[key]))
                .join("&") :
            "";
        let linkHeaders = ''
        let userRepoIssues = `${this.state.user}/${this.state.repo}/issues`
        let fullUrl = `${baseUrl}/${userRepoIssues}${params}${this.state.page}`
        fetch(fullUrl, headers)
            .then(response => {
                if (response.headers.get('Link')) {
                    linkHeaders = this.parseLinkHeader(response.headers.get('Link'))
                }
                if (response.ok) return response.json();
                throw new Error('Request failed');
            })
            .then(data => {
                this.setState({
                    pages: linkHeaders,
                    issues: data,
                    loading: false,
                    error: null,
                    showBody: {},
                });
            })
            .catch(error => {
                this.setState({
                    loading: false,
                    error: error
                });
            });
    }
    handlePageChange(page) {
        const newPage = (previousState) => {
            return { ...previousState, page: page, currentPageNumber: +page.split('=').pop(), loading: true };
        };
        this.setState(newPage, () => {
            this.fetchAllIssues()
        })
    }
    handleFilterChange(toFilter) {
        let newObj = { ...this.state.listFilter, ...toFilter }
        const newFilter = (previousState) => {
            let stateTest = previousState.listFilter.state !== newObj.state ? true : false
            return { ...previousState, listFilter: newObj, loading: stateTest };
        };
        this.setState(newFilter, () => {
            Object.keys(toFilter)[0] === "state" && this.fetchAllIssues()
        })
    }
    parseLinkHeader = (header) => {
        if (header.length === 0) {
            throw new Error("input must not be of zero length");
        }
        return header.split(/(?!\B"[^"]*),(?![^"]*"\B)/).reduce((links, part) => {
            const section = part.split(/(?!\B"[^"]*);(?![^"]*"\B)/);
            if (section.length < 2) {
                throw new Error("section could not be split on ';'");
            }
            const url = section[0].replace(/<(.*)>/, '$1').trim();
            const name = section[1].replace(/rel="(.*)"/, '$1').trim();
            links[name] = url;

            return links;
        }, {});
    }
    handleSortingChange() {
        let newObj = { ...this.state.sort };
        const newState = (previousState) => {
            let stateTest = previousState.sort !== newObj ? true : false
            return { ...previousState, sort: !previousState.sort, loading: stateTest, sortName: !previousState.sort ? "Latest" : "Oldest" };
        };
        this.setState(newState, () => {
            this.fetchAllIssues()
        })
    }
    showBodyHandler = (obj) => {
        this.setState({ showBody: obj })
    }
    handleDateChange (date)  {
        const newFilter = (previousState) => {
            let stateTest = previousState.since !== date ? true : false
            return { ...previousState, since: date, loading: stateTest };
        };
        this.setState(newFilter, () => {
            this.fetchAllIssues()
        })
    }
    render() {
        return (
            <div className="container">
                <nav className="panel" >
                    <IssuesFiltering
                        listFilter={this.state.listFilter}
                        onFilterChange={(event) => this.handleFilterChange(event)}
                        sortName={this.state.sortName}
                        clicked={(event) => this.handleSortingChange(event)}
                        startDate={this.state.since}
                        handleDateChange={(event)=>this.handleDateChange(event)}
                    />
                    {this.state.loading ?
                        <Spinner />
                        : <GitHubIssue state={this.state} showBodyHandler={this.showBodyHandler} />}
                    <nav className="navbar  is-fixed-bottom is-transparent">
                        <div className="navbar-start" style={{ flexGrow: 1, justifyContent: "center" }}>
                            <div className="navbar-item">
                                <Pagination pages={this.state.pages} currentPageNumber={this.state.currentPageNumber} onPageChange={(event) => this.handlePageChange(event)} />
                            </div>
                        </div>
                    </nav>
                </nav></div>);
    }
}
export default GithubIssues;

