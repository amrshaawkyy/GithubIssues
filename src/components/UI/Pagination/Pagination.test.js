import React from 'react';
import { configure, shallow , mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Pagination from './Pagination';
import GithubIssues from '../../../containers/GithubIssues/GithubIssues';
import expect from 'expect';
configure({ adapter: new Adapter() });

describe('<IssuesFiltering />', () => {
    let wrapper;
    afterEach(() => {
        expect(wrapper).toMatchSnapshot();
    });
    it("OnClick pageNext change  currentPageNumber to 2",()=>{
        const comp=mount( <GithubIssues/>);
        const PaginationComp = comp.find(Pagination);
        const button=PaginationComp.find('button').at(2);
        button.simulate('click');
        expect(comp.state().currentPageNumber).toEqual(2);
    });
});
