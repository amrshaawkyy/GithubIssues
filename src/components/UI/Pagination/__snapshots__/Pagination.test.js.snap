// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IssuesFiltering /> OnClick Change State SortName to Oldest 1`] = `undefined`;

exports[`<IssuesFiltering /> OnClick pageNext change  currentPageNumber to 2 1`] = `undefined`;
